camayoc/tests/qcs/test_scan.py

```diff
@@ -23,63 +23,100 @@
 
 
 def sources():
-    """Gather sources from config file."""
+    """Gather sources from config file.
+
+    If no config file is found, or no sources defined,
+    then an empty list will be returned.
+
+    If a test is parametrized on the sources in the config file, it will skip
+    if no sources are found.
+    """
     try:
-        srcs = config.get_config()['qcs']['profiles']
+        srcs = config.get_config()['qcs']['sources']
     except (ConfigFileNotFoundError, KeyError):
         srcs = []
     return srcs
 
 
-def wait_until_complete(scan, timeout=120):
-    """Wait until the scan has either failed or completed.
+def first_source():
+    """Gather sources from config file.
+
+    If no source is found in the config file, or no config file is found, a
+    default source will be returned.
+    """
+    try:
+        src = [config.get_config()['qcs']['sources'][0]]
+    except (ConfigFileNotFoundError, KeyError):
+        src = [{'hosts': ['localhost'], 'name':'localhost', 'auths':'default'}]
+    return src
+
+
+def wait_until_state(scan, timeout=120, state='completed'):
+    """Wait until the scan has failed or reached desired state.
+
+    The default state is 'completed'.
 
     This method should not be called on scan jobs that have not yet been
     created, are paused, or are canceled.
 
     The default timeout is set to 120 seconds, but can be overridden if it is
     anticipated that a scan may take longer to complete.
     """
-    if scan.status() in ['paused', 'cancelled']:
-        raise UserWarning(
-            'You have called wait_until_complete() on a scan '
-            'that is paused or cancelled, and this could hang forever.'
-            ' This exception has been raised instead.'
-        )
     while (
         not scan.status() or not scan.status() in [
             'failed',
-            'completed']) and timeout > 0:
+            state]) and timeout > 0:
         time.sleep(5)
         timeout -= 5
+    if timeout <= 0:
+        raise UserWarning(
+            'You have called wait_until_state() on a scan '
+            'and the scan timed out while waiting to achieve the state="{}"'
+            ' When the scan timed out, it had the state="{}".'
+            .format(state, scan.status())
+        )
```

Comment:
This now raises an exception if it times out waiting for a state. This now allows the function to be used for any state.

Do you think UserWarning is appropriate or do you have another suggestion?

Reply:
I think this if statement could be moved inside the while loop and raise an exception instead of a warning. The case I see is this, imagine that we left the user warning when developing the test and the test is passing (false positive), in this case we will never be able to capture that. But if an exception is raised that will generate an error on the test results.

I think it is better to be (or feel) safe when running on jenkins by being noisy.

You can create a custom exception for this case, maybe a TimeoutError exception?

Reply:
Done, and TimeoutError is a built in!

Artificed a scenario for it to time out, here is an example output.

```
E               TimeoutError: You have called wait_until_state() on a scan
E               and the scan timed out while waiting to achieve the state="completed"
E               When the scan timed out, it had the state="created".
```

Reply:
That is meant to be used for OS related operations. Better to create one inside the exceptions module.

> Raised when a system function timed out at the system level. Corresponds to errno ETIMEDOUT.

Reply:
OK that sounds good, will do